### ISSUE TITLE
Add ability to easily add a Framework Reference to an SDK-style project

### ISSUE THREAD

**Issue by DamianEdwards** [dotnet/project-system#8558]
We often get questions from customers asking how to use ASP.NET Core types in class libraries, generally when they're trying to create reusable libraries of web-related code for their ASP.NET Core apps.

Discovering that one needs to add a framework reference to Microsoft.AspNetCore.App to the project is fairly difficult from within the product today, although [there are docs](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/target-aspnetcore?view=aspnetcore-6.0&tabs=visual-studio). One thing we can do to perhaps make this a smidge easier is make it simple to add Framework References from the Solution Explorer Dependencies node:

<img width="176" alt="image" src="https://user-images.githubusercontent.com/249088/193369991-91b0c169-3a4d-4ade-8d26-d6ca89051090.png">

<img width="306" alt="image" src="https://user-images.githubusercontent.com/249088/193370141-772e9f26-797e-405a-9a6a-6e39abfa0bfe.png">

I'm envisaging a new context menu item there labelled "Add a Framework Reference..." that opens the add reference dialog on a new pane that allows selecting from a list of discovered shared frameworks (ref packs) for the version of .NET targeted by the project.

As an extension, having the editor be able to resolve types automatically from shared frameworks that aren't referenced and then add the required framework reference automatically (like is supported for types in unimported namespaces and unreferenced NuGet packages today) would be a hugely helpful addition.

**Comment by drewnoakes** [comment]
> selecting from a list of discovered shared frameworks (ref packs) for the version of .NET targeted by the project

This sounds like the main challenge. Do you know how this would work?

> having the editor be able to resolve types automatically from shared frameworks that aren't referenced and then add the required framework reference automatically

This would be great, but likely driven from the Roslyn side. Can you file the request against Roslyn and tag it Area-IDE?

**Comment by DamianEdwards** [comment]
> > selecting from a list of discovered shared frameworks (ref packs) for the version of .NET targeted by the project

> This sounds like the main challenge. Do you know how this would work?

I think perhaps this should likely be resolved via metadata we have in the SDK via MSBuild, similar to the way that the SDK knows what the ref pack and runtime pack versions are for any framework references based on the TFM. So perhaps a new target (if there isn't one already) that resolves the available frameworks given the TFM? @marcpopMSFT @dsplaisted for thoughts.

**Comment by tmeschter** [comment]
And the result would be adding something like the following to the project file?

```xml
<ItemGroup>
  <FrameworkReference Include="Microsoft.AspNetCore.App" />
</ItemGroup>
```

**Comment by DamianEdwards** [comment]
@tmeschter yes, although it's dependent on the shared framework they want to add a reference to. Luckily the matrix isn't hugely complex here as we only ship 3 shared frameworks today:

- Microsoft.NETCore.App (added by default to all SDK-style projects so shouldn't need to be offered as a choice)
- Microsoft.AspNetCore.App (added by default to projects using the Microsoft.NETCore.Sdk.Web SDK)
- Microsoft.WindowsDesktop.App (added by default to projects using the Microsoft.NETCore.Sdk SDK **and** specifying the `UseWinForms` or `UseWpf` properties)

We should confirm with the right folks what the right project modification would be for something like a class library that wants to use something from Windows Forms, i.e. should it just add `<FrameworkReference Include="Microsoft.WindowsDesktop.App" />` or should it set `<UseWinForms>true</UseWinForms>`?

**Comment by DamianEdwards** [comment]
> It probably wouldn't be too hard to refactor the code to add a task/target to get the available framework references.

@dsplaisted we'd likely want this to only return available **additional** framework references, i.e. those that aren't already referenced.

**Comment by kvenkatrajan** [comment]
Coordination with sdk team to add an extension point for determining the frameworks available to be added

**Comment by tmeschter** [comment]
After further discussion in the MSBuild/CLI/partners sync up I'm sending back for re-triage.

Some relevant points from the meeting:

1. ASP.NET projects get the right FrameworkReference via the Web SDK.
2. WinForms and WPF get the right FrameworkReferences via the `UseWinForms` and `UseWPF` properties.
3. We could have an "Add Framework Reference" action (similar to "Add Project Reference" or "Add COM Reference") that opens the Add Reference dialog to a new "Framework" section. Adding the ASP.NET framework would need to add an actual `FrameworkReference` item, while adding WinForms/WPF would set the `UseWinForms`/`UseWPF` properties instead.
4. The SDK would need to tell us what Frameworks are available, but how we translate that information into the above actions is something we would need to design if we decided to move forward with this.
5. We would also want corresponding Roslyn Code Actions that would do something similar directly from code (e.g. if you try to use a type named "Button" there would be code actions to "turn on" WinForms or WPF").

**Comment by DamianEdwards** [comment]
> 5. We would also want corresponding Roslyn Code Actions that would do something similar directly from code (e.g. if you try to use a type named "Button" there would be code actions to "turn on" WinForms or WPF").

https://github.com/dotnet/roslyn/issues/64505
